h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -437,6 +440,132 @@ public void modelFailed(Model failedModel) {
     }
   } // class CartesianWalker
 
+
+  /**
+   * Random iteration over the hyper-parameter space.  Does not repeat previously-visited combinations. 
+   * Implementation differs from {@link RandomDiscreteValueWalker} as we first eagerly evaluate the whole space, shuffle it and then iterate sequentially.
+   * Materialised grid might be taking considerable amount of memory so it needs to be taken into account.
+   */
+  public static class MaterializedRandomWalker<MP extends Model.Parameters>
+          extends BaseWalker<MP, HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria> {
+
+    Random random;
+    ArrayList<Function<HashMap<String, Object>, HashMap<String, Object>>> _filterFunctions;
```

Comment:
does it have to be `HashMap`? wouldn't work if the filter only manipulates `Map`?

Reply:
I will think on the best representation of the grid item as we will need to consider memory impact.

Reply:
you should definitely use interfaces here, you can weight in memory/performace when actualy instantiating the values